pkg/ivd/ivd_protected_entity_type_manager_util.go

```diff
@@ -0,0 +1,204 @@
+		return err
+	}
+	sEnc := string(secret.Data["csi-vsphere.conf"])
+	lines := strings.Split(sEnc, "\n")
+
+	for _, line := range lines {
+		if strings.Contains(line, "VirtualCenter") {
+			parts := strings.Split(line, "\"")
+			params["VirtualCenter"] = parts[1]
+		} else if strings.Contains(line, "=") {
+			parts := strings.Split(line, "=")
+			key := strings.TrimSpace(parts[0])
+			value := strings.TrimSpace(parts[1])
+			params[key] = value[1 : len(value)-1]
+		}
+	}
+
+	return nil
+}
+
+func createCnsVolumeWithClusterConfig(ctx context.Context, config *rest.Config, client *govmomi.Client, md Metadata, logger logrus.FieldLogger) (string, error) {
+	params := make(map[string]interface{})
+	err := retrievePlatformInfoFromConfig(config, params, logger)
+	if err != nil {
+		logger.WithError(err).Errorf("Failed to retrieve VC config secret: %+v", err)
+		return "", err
+	}
+
+	clusterId, ok := params["cluster-id"].(string)
+	if !ok {
+		logger.WithError(err).Errorf("Failed to retrieve cluster id")
+		return "", err
+	}
+
+	user, ok := params["user"].(string)
+	if !ok {
+		logger.WithError(err).Errorf("Failed to retrieve vsphere user")
+		return "", err
+	}
+	logger.Debugf("Retrieved cluster id, %v, and vSphere user, %v", clusterId, user)
+
+	// Preparing for the VolumeCreateSpec for the volume provisioning
+	logger.Debug("Preparing for the VolumeCreateSpec for the volume provisioning")
+	dsList, err := findAllDatastores(ctx, client.Client)
+	if err != nil {
+		logger.WithError(err).Error("Failed to find any datastore in the underlying vSphere")
+		return "", err
+	}
+
+	var metadataList []cnstypes.BaseCnsEntityMetadata
+	metadata := &cnstypes.CnsKubernetesEntityMetadata{
+		CnsEntityMetadata: cnstypes.CnsEntityMetadata{
+			EntityName:  md.VirtualStorageObject.Config.Name,
+			Labels:      md.ExtendedMetadata,
+		},
+		EntityType: string(cnstypes.CnsKubernetesEntityTypePV),
+	}
+	metadataList = append(metadataList, cnstypes.BaseCnsEntityMetadata(metadata))
+
+	var cnsVolumeCreateSpecList []cnstypes.CnsVolumeCreateSpec
+	cnsVolumeCreateSpec := cnstypes.CnsVolumeCreateSpec{
+		Name:        md.VirtualStorageObject.Config.Name,
+		VolumeType: string(cnstypes.CnsVolumeTypeBlock),
+		Datastores: dsList,
+		Metadata: cnstypes.CnsVolumeMetadata{
+			ContainerCluster: cnstypes.CnsContainerCluster{
+				ClusterType: string(cnstypes.CnsClusterTypeKubernetes), // hard coded for the moment
+				ClusterId:   clusterId,
+				VSphereUser: user,
+			},
+			EntityMetadata: metadataList,
+		},
+		BackingObjectDetails: &cnstypes.CnsBlockBackingDetails{
+			CnsBackingObjectDetails: cnstypes.CnsBackingObjectDetails{
+				CapacityInMb: md.VirtualStorageObject.Config.CapacityInMB,
+			},
+		},
+	}
+
+	cnsVolumeCreateSpecList = append(cnsVolumeCreateSpecList, cnsVolumeCreateSpec)
+	logger.Debugf("Provisioning volume using the spec: %v", cnsVolumeCreateSpec)
+
+	// provision volume using CNS API
+	cnsClient, err := cns.NewClient(ctx, client.Client)
```

Comment:
Do we want to always create a new CNSClient here?  Does this need to closed/released?

Reply:
You are right. It is better to create one along with initiating a new IVD PETM. Will change it.